### PR TITLE
x64: Handle Intel JCC Erratum

### PR DESCRIPTION
**Pull request by merryhime**
Also see [relevant white paper](https://www.intel.com/content/www/us/en/support/articles/000055650/processors.html).

This implementation adds NOP padding before all affected JCC/JMP/CALL/RET instructions.

~~**Please note:** This implementation will split macro-op fusable op-JCC pairs. An improved implementation would add padding before such pairs, instead of in between these instructions.~~

### CODE REVIEW

**Comment by BhaaLseN** [comment]
Would you mind adding a TL;DR comment somewhere, briefly describing what the problem is there (and how to address it)?

**Comment by MerryMage** [comment]
Added a comment.

File: Source/Core/Common/x64Emitter.cpp [hunk 1]
```diff
@@ -434,35 +453,67 @@ void XEmitter::JMP(const u8* addr, bool force5Bytes)
...
+    arg.WriteRest(this);
+  };
+
+  if (!cpu_info.bJccErratum)
+  {
+    emit();
+    return;
+  }
+
+  u8* instructon_begin = code;
```
Comment (CookiePLMonster) [inline Source/Core/Common/x64Emitter.cpp]: Typos?

**Comment by MerryMage** [comment]
Added handling of macrofused instruction pairs.

Todo:
- [ ] Audit all `J_CC` locations to ensure no invalid assumed fusions occur.

File: Source/Core/Common/x64Emitter.cpp [hunk 2]
```diff
@@ -1683,11 +1754,13 @@ void XEmitter::CMP_or_TEST(int bits, const OpArg& a1, const OpArg& a2)
   CheckFlags();
   if (a1.IsSimpleReg() && a2.IsZero())  // turn 'CMP reg, 0' into shorter 'TEST reg, reg'
   {
-    WriteNormalOp(bits, NormalOp::TEST, a1, a1);
+    EmitJccFusableInstruction(JccFusionType::All,
+                              [&] { WriteNormalOp(bits, NormalOp::TEST, a1, a2); });
```
Comment (Sintendo) [inline Source/Core/Common/x64Emitter.cpp]: Copy-paste error, last parameter should be `a1`.

File: Source/Core/Common/x64Emitter.h [hunk 3]
```diff
@@ -430,6 +430,14 @@ class XEmitter
   void PUSHF();
   void POPF();
 
+  // Intel CPU microcode migitations that fix Intel erratum SKX102 affect the caching of certain
+  // jmp/call/ret instructions: If such instructions (or such macrofused instruction pairs) span a
+  // 32-byte boundary, they will not be cached in the host CPU's icache. In order to reduce the
```
Comment (phire) [inline Source/Core/Common/x64Emitter.h]: I believe they are still cached in L1i cache, but not in the µOP cache (sometime called the L0i cache)

File: Source/Core/Core/PowerPC/Jit64/JitAsm.cpp [hunk 4]
```diff
@@ -81,6 +81,7 @@ void Jit64AsmRoutineManager::Generate()
   ResetStack(*this);
 
   SUB(32, PPCSTATE(downcount), R(RSCRATCH2));
+  NOP();  // Prevent op-Jcc fusion.
```
Comment (phire) [inline Source/Core/Core/PowerPC/Jit64/JitAsm.cpp]: Why are we preventing op-Jcc fusion? I though the emitter now handles this.

Comment (MerryMage) [reply Source/Core/Core/PowerPC/Jit64/JitAsm.cpp]: Here, the later JLE in `dispatcher` will be detected as fusable with the SUB, and thus the SUB will be memmoved. However, this invalidates the value in `dispatcher`.

An alternative way of preventing this is to modify `GetCodePtr` as so:

```
const u8* XEmitter::GetCodePtr() const
{
  jcc_fusion_type = JccFusionType::None;
  return code;
}
```

i.e. prevent the possibility of `memmove`-ing code across such locations

**Comment by MayImilae** [comment]
@JMC47 please do thorough performance testing! We'll need a lot of data for when this hits a Progress Report.

File: Source/Core/Core/PowerPC/Jit64/JitAsm.cpp (hunk 4 above)
Comment (phire) [reply Source/Core/Core/PowerPC/Jit64/JitAsm.cpp]: It's fine, but would be nice if the comment elaborated.

File: Source/Core/Common/x64Emitter.cpp (hunk 2 above)
Comment (MerryMage) [reply Source/Core/Common/x64Emitter.cpp]: Thanks.

File: Source/Core/Core/PowerPC/Jit64/JitAsm.cpp (hunk 4 above)
Comment (MerryMage) [reply Source/Core/Core/PowerPC/Jit64/JitAsm.cpp]: Done.

**Comment by Rumi-Larry** [comment]
What happened to this pr?

**Comment by Rumi-Larry** [comment]
**Poke**

**Comment by Rumi-Larry** [comment]
> 
> 
> **Poke**

I'm just curious why this is no longer necessary.

**Comment by Rumi-Larry** [comment]
I'm really curious, that's all. Please someone explain why this wasn't pursued.

**Comment by MerryMage** [comment]
Realistically this affects only a subset of processors that are 10 years old now. I'm happy to rebase if someone wishes to merge this, but I don't really mind either way.

**Comment by Rumi-Larry** [comment]
> 
> 
> Realistically this affects only a subset of processors that are 10 years old now. I'm happy to rebase if someone wishes to merge this, but I don't really mind either way.

Thank you!

**Comment by mirh** [comment]
Shouldn't this affect skylake (2015) to comet lake (at least the early steppings, 2019)?

**Comment by mirh** [comment]
Still no real benchmark to speak off, anyway there's also an automatic build switch to enable this now. 
https://github.com/yuzu-emu/yuzu/pull/7007

**Comment by merryhime** [comment]
> Still no real benchmark to speak off, anyway there's also an automatic build switch to enable this now. [yuzu-emu/yuzu#7007](https://github.com/yuzu-emu/yuzu/pull/7007)

This PR corrects emission in the JIT; this won't be fixed by a compiler flag.

**Pull request closed without merging**